public/js/components/Editor.js

```diff
@@ -174,25 +176,62 @@ const Editor = React.createClass({
   componentWillReceiveProps(nextProps) {
     // This lifecycle method is responsible for updating the editor
     // text.
-    const sourceText = nextProps.sourceText;
+    const { sourceText, selectedLocation } = nextProps;
 
     if (!sourceText) {
-      this.setText("");
-      this.editor.setMode({ name: "text" });
-    } else if (!isTextForSource(sourceText)) {
-      // There are only 2 possible states: errored or loading. Do
-      // nothing except put a message in the editor.
-      this.setText(sourceText.get("error") || "Loading...");
-      this.editor.setMode({ name: "text" });
-    } else if (this.props.sourceText !== sourceText) {
-      // Only update it if the `sourceText` object has actually changed.
-      // It is immutable so it will always change when updated.
-      this.setText(sourceText.get("text"));
-      this.setMode(sourceText);
```

Comment:
Personally, I really really liked the terse-ness of the previous code so I'm a little sad to see it go. There's a bunch of things going on here so it's really nice to see it all in just 10 lines of code and it's easy to follow. I don't think we need to do anything with documents for loading/error accept make sure a fresh one exists, so you could reduce bring back the old code and just make a `this.freshDocument()` function and call that before loading in the error/loading text.

For example, if the sourceText does not exist we do need to reset the editor to `""`, I think because on navigation it gets into that state sometimes and we need to make sure it's not showing the text from before (I can't remember the exact step to reproduce this)

Pulling out the loading the actual source text makes sense though. I'm not sure if `replaceDocument` is cheap if the current doc is the same; it probably is, but if not we should also keep the cheap `this.props.sourceText !== sourceText` check.

I need to head out right now but I can provide some sample code to explain more clearly what I'm suggesting tomorrow morning (or maybe later tonight)